### ISSUE TITLE
Add an API to execute a console program and capture its output in a simple way

### ISSUE THREAD

**Issue by GSPP** [dotnet/corefx#3483]
Clearly, we can use the `Process` class to run a program, redirect its output and capture the output. But doing so correctly is unfathomably complicated:

1. You need to read both StdOut and StdErr. Failing to do so can result in hangs.
2. Output events can arrive *after* WaitForExit has returned. (Update: This concern does not exist.)
3. Output events signal their completing by sending one `null` string at the end.
3. `myProcess.StandardOutput.ReadAsync` throws an exception for the last read with a confusing message (http://stackoverflow.com/questions/32800802/unable-to-read-output-from-child-process-from-c-sharp-app-using-async-methods#32800802).
4. Reading binary output requires contortions (`myProcess.StandardOutput.BaseStream`) where it is not clear whether this is supported at all. We are reaching into the `BaseStream` property which might expose an internal object not meant for consumption.
5. Failing to quickly read the redirected streams stalls the child process.
6. Many more on this list: http://csharptest.net/321/how-to-use-systemdiagnosticsprocess-correctly/
7. Probably, even more that are on nobodies list.

This API is terrible!

Please provide an easy and reliable way to capture output. Here's a strawman for a simple API

`IEnumerable<string> Process::StartWithOutputRedirected()`

Now, this clearly is insufficient and probably not nice API design but you get the point.

The following goals should be achieved:

1. Little potential for mistakes.
2. Streaming output for unbounded output size.
3. Text and binary output.
4. Differentiate between StdOut and StdErr. For example by returning an `IEnumerable` of a DTO like `struct ConsoleOutput { string Data; ConsoleOutputType Type; /*Output, Error*/ }`.
5. The output stream should end at exactly the same point in time that WaitForExit ends at. Delay both to only complete when the other also has completed.
6. The API should allow reading in a non-linewise fashion.
7. A simplified version that just captures all output on both channels into one `string`. This is probably what most consumers want anyway.

**Comment by pallavit** [comment]
@gspp Did you consider using OutputDataRecieved and OutputErrorRecieved which lets you read these streams asynchronously?

**Comment by GSPP** [comment]
@pallavit yes, I did. Search this thread for "event" to find some perils with using them.

**Comment by GSPP** [comment]
Here's a little API concept for this:

```C#
class Process
{
    //Only available on Process instances that were returned from Start
    //and that have redirected output. Otherise null or throwing.
    public ProcessOutput ProcessOutput { get; private set; }

    //Add a better way to wait for exit which is a common requirement.
    public Task<int> WaitForExitAsync(CancellationToken ct) { throw new NotImplementedException(); }

    //Convenience method.
    public static ProcessOutput StartWithOutput(ProcessStartInfo processStartInfo) { throw new NotImplementedException(); }
}

class ProcessOutput
{
    //The idea is that callers can obtain process output in any convenient format they like.
    //In order to allow for streaming each of the two outputs can only be obtained once.
    //Trying to obtain it a second time throws.
    //This API easily allows to combine StdOut and StdErr
    //into one string/stream which is a common requirement.
    //It also allows for separate processing.
    //Probably, each of these should have an overload without arguments
    //that simply includes both streams.

    public string GetStandardOutputAsString(bool includeStdOut, bool includeStdErr) { throw new NotImplementedException(); }
    public IEnumerable<string> GetStandardOutputAsLines(bool includeStdOut, bool includeStdErr) { throw new NotImplementedException(); }
    public Stream GetStandardOutputAsStream(bool includeStdOut, bool includeStdErr) { throw new NotImplementedException(); }
    public byte[] GetStandardOutputAsBytes(ProcessOutputType processOutputType) { throw new NotImplementedException(); }

    //This class could also have an async version for the string and lines versions.

    //This provides the lines in one combined sequence but with a way to discern the source.
    public IEnumerable<ProcessOutputLine> GetStandardOutputAsLinesCombined() { throw new NotImplementedException(); }
}

class ProcessOutputLine
{
    public ProcessOutputType Type { get; set; }
    public string Data { get; set; }
}

enum ProcessOutputType
{
    StdOut,
    StdErr,
}
```

The idea behind this API design is to not pollute the `Process` class with all those methods. At the same time it is very easy to obtain process output in a convenient way.

**Comment by pallavit** [comment]
> Output events can arrive after WaitForExit has returned.

Do you mean when calling WaitForExit(int) or WaitForExit() method. If it is the formal, the best practice is to call the WaitForExit() after recieving true from WaitForExit(int). If you see this behavior with WaitForExit() can you share a code sample that repros this on NetCore?

**Comment by GSPP** [comment]
@pallavit it is WaitForExit. I don't have code handy but it is clear that the issue exists:

1. WaitForExit simply waits on the handle.
2. Events are delivered on a different thread. So an event could be delivered and the OS suspends the event thread on the first instruction of the handler. Then, the process exits and WaitForExit returns. Then, the event is actually processed.

The proposed behavior is to delay WaitForExit from returning until all data has been processed by all event handlers. This is what, I think, all apps want and really almost none enforce (=> latent bugs).

**Comment by pallavit** [comment]
Sorry, I am not sure I understand the issue.

>  I don't have code handy but it is clear that the issue exists:
1. WaitForExit simply waits on the handle.
2. Events are delivered on a different thread. So an event could be delivered and the OS suspends the event thread on the first instruction of the handler. Then, the process exits and WaitForExit returns. Then, the event is actually processed.

Events are fired on the same thread on which they are raised. Process.WaitForExit internally calls WaitForExitCore which in its finally block calls WaitUntilEOF on the async stream reader which awaits a task that is responsible for reading the buffer, flushing the message queue and calling the event handler per line read(all these steps will run on the same thread on which ReadBufferAsync is called). So the Process.WaitForExit() won't return until AsyncStreamReader.WaitUntilEOF returns which will await the above task.

It is possible that there are bugs lurking around but please share a repro if you see this behavior as it is unexpected and I would like to take a closer look at it.

**Comment by GSPP** [comment]
OK, I did not see the call to WaitUntilEOF and that makes one of the scenarios that I thought was possible impossible.

Did this change over time? If not then I probably misremembered. In any case this part of the issue is no longer relevant.

Just to make sure I tried to repro and could not. Posting the code here for reference:

```
    class Program
    {
        static void Main(string[] args)
        {
            RunOnce();

            ThreadPool.SetMinThreads(1000, 1000);

            ParallelEnumerable.Range(0, int.MaxValue)
                .WithDegreeOfParallelism(64)
                .ForAll(_ => RunOnce());
        }

        static void RunOnce()
        {
            var p = Process.Start(new ProcessStartInfo()
            {
                FileName = "cmd.exe",
                Arguments = "/c echo 1234&echo.& echo 1234",
                WorkingDirectory = null,
                CreateNoWindow = true,
                WindowStyle = ProcessWindowStyle.Hidden,
                UseShellExecute = false,
                RedirectStandardOutput = true,
                RedirectStandardError = true,
            });

            var sb = new StringBuilder();
            using (p)
            {
                int completionCount = 0;

                p.OutputDataReceived += (object sender, DataReceivedEventArgs e) =>
                {
                    if (e.Data != null)
                        sb.AppendLine(e.Data);
                    else
                        Interlocked.Increment(ref completionCount);
                };
                p.ErrorDataReceived += (object sender, DataReceivedEventArgs e) =>
                {
                    if (e.Data != null)
                        sb.AppendLine(e.Data);
                    else
                        Interlocked.Increment(ref completionCount);
                };

                p.BeginOutputReadLine();
                p.BeginErrorReadLine();

                p.WaitForExit();

                Debug.Assert(Volatile.Read(ref completionCount) == 2);
            }

            var output = sb.ToString();
            Debug.Assert(output == "1234\r\n\r\n1234\r\n");
        }
    }
```

Still, the other API usability issues stand.

**Comment by GSPP** [comment]
Btw, in this benchmark there is a serious scalability limitation which is the global lock in `StartWithCreateProcess`. Of 64 threads 62 were stuck there when I paused the debugger.

![process lock](https://cloud.githubusercontent.com/assets/12032350/11559852/692035d6-99bc-11e5-8071-1fd66d517ba5.PNG)

Is this by design? Seems like a scalability problem with many cores and process heavy workloads.

**Comment by pallavit** [comment]
> Seems like a scalability problem with many cores and process heavy workloads.

Honestly, I am not sure how much a problem will this be in a practical setting. You are the first one to raise the issue :). 

> Just to make sure I tried to repro and could not.

Great ! I hope async events solve most of the issues that you raised with reading sync streams, Let me know if you still see issues or else I will go ahead and close this.

**Comment by GSPP** [comment]
I did not open this issue to get a particular bug fixed. My observation is that the Process output redirection API is hard to use in general. I have struggled with it, so have others (https://www.google.com/webhp?complete=1&hl=en&gws_rd=ssl#complete=1&hl=en&q=site:stackoverflow.com+.net+Process+redirect 32,000 results).

My main suggestion is to create an easier API for this common case. I'm not concerned with any particular issue. I think it's appropriate to close the issue if the team decides that it will not pursue this API improvement any further. Has this been decided?

So far we have resolved one of approximately 8 concerns (number 2).

**Comment by pallavit** [comment]
>  My observation is that the Process output redirection API is hard to use in general. 

I completely agree with you that the sync APIs are can be difficult to use. However you mentioned two issues with respect to ouput events namely 
> 1. Output events can arrive after WaitForExit has returned. (Update: This concern does not exist.)

Which is a non-issue.

> 2. Output events signal their completing by sending one null string at the end.

This was mainly added for EOF and I would not really look at it as a bug.

If you know of any other issues with output events, I would like to know about them. I do like your suggestion about WaitForExitAsync. I don't think we can add it to the current Process class right now, however, I will definitely keep it in mind for future.

As for the rest of the APIs you mentioned, I don't think we can add them to the current process design for two reasons
1. Process o/p events seem to address most of the issues you raised.
2. It pollutes the current design. We already have two ways of reading streams, which we can't remove and adding the third one complicates the design further. However, I will definitely keep your suggestions in mind, in case any opportunity arises to update this in future.

**Comment by pallavit** [comment]
Thanks again for the feedback. I will keep your feedback in mind in case an opportunity arises in future.

**Comment by GSPP** [comment]
Here's a typical case for what it's like to use the Process class and receive child output: http://stackoverflow.com/questions/35184529/why-does-process-waitforexit-sometimes-hang-though-called-upon-an-exited-proce

fyi.
